providers/implementations/keymgmt/dsa_kmgmt.c

```diff
@@ -378,8 +356,197 @@ static int dsa_validate(void *keydata, int selection)
     return ok;
 }
 
+static void *dsa_gen_init(void *provctx, int selection)
+{
+    OPENSSL_CTX *libctx = PROV_LIBRARY_CONTEXT_OF(provctx);
+    struct dsa_gen_ctx *gctx = NULL;
+
+    if ((selection & (OSSL_KEYMGMT_SELECT_KEYPAIR
+                      | OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS)) == 0)
+        return NULL;
+
+    if ((gctx = OPENSSL_zalloc(sizeof(*gctx))) != NULL) {
+        gctx->libctx = libctx;
+        gctx->pbits = 2048;
+        gctx->qbits = 224;
+        gctx->md = NULL;
+        gctx->type = DSA_PARAMGEN_TYPE_FIPS_186_4;
+        gctx->gindex = -1;
+        if ((selection & OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS) != 0)
+            gctx->param_gen = 1;
```

Comment:
I would actually simply store `selection` in the context, then deal with its value in `dsa_gen()`.  Also, as an option, I'd deal with the possibility that you could get a selection that covers everything, i.e. that someone some day could call with `OSSL_KEYMGMT_SELECT_ALL`, even though that's not *currently* happening.

Reply:
good idea..
On that note I will copy the template values first.. and then allow the other values to override the params (seed and gindex) afterwards..